include/dqlite.h

```diff
@@ -17,8 +17,278 @@
+ * string, as passed to dqlite_server_create or dqlite_server_set_peer. @fd is
+ * an address where a socket representing the connection should be stored. The
+ * callback should return zero if a connection was established successfully or
+ * nonzero if the attempt failed.
+ */
+typedef int (*dqlite_connect_func)(void *arg, const char *addr, int *fd);
+
+/* The following functions (dqlite_server_* and the functions that mimic the
+ * sqlite3 API) use SQLite result codes. dqlite tries to return result codes
+ * that match what SQLite would return in the same situation, but in some cases
+ * will return the less specific SQLITE_ERROR instead. Certain dqlite-specific
+ * error conditions have their own codes that are compatible with the SQLite
+ * set: */
+
+#define SQLITE_IOERR_NOT_LEADER (SQLITE_IOERR | (40 << 8))
+#define SQLITE_IOERR_LEADERSHIP_LOST (SQLITE_IOERR | (41 << 8))
+
+/* When some version of dqlite returns SQLITE_ERROR under certain
+ * circumstances, the developers reserve the right to switch to a more specific
+ * error code in future releases without announcing a breaking change. */
+
+/**
+ * Start configuring a dqlite server.
+ *
+ * The server will not start running until dqlite_server_start is called. @path
+ * is the path to a directory where the server (and attached client) will store
+ * its persistent state; the directory must exist. @addr is the abstract
+ * address of the server. A pointer to the new server object is stored in
+ * @server on success.
+ *
+ * The abstract address will be passed to the custom connect function configured
+ * with dqlite_server_set_connect_func, or to the default connect function if no
+ * custom function is configured. The default connect function accepts IPv4
+ * addressed, IPv6 addresses, and Unix-domain socket addresses; for a complete
+ * description of the syntax, see dqlite_server_set_bind_address.
+ *
+ * No reference to @path or @addr is kept after this function returns.
+ */
+int dqlite_server_create(const char *path,
+			 const char *addr,
+			 dqlite_server **server);
+
+/**
+ * Declare the abstract address of a server that will be responsible for
+ * bootstrapping the cluster.
+ *
+ * The "bootstrap" node must be the first node to start up, and it will
+ * automatically assume leadership for the cluster's first term. To configure
+ * @server to act as the bootstrap server itself, pass @server's own abstract
+ * address to this function. At most one bootstrap server may be specified. The
+ * bootstrap server declaration only has an effect when a server is starting up
+ * for the first time (but it is harmless to call this function
+ * unconditionally).
+ *
+ * No reference to @addr is kept after this function returns.
+ */
+int dqlite_server_set_bootstrap_address(dqlite_server *server,
+					const char *addr);
+
+/**
+ * Declare the abstract address of another server that will participate in a
+ * cluster with this server.
```

Comment:
As far as I can tell the purpose of this function is to let a new server that is starting for the first time to join an existing cluster. So I'd change this docstring to:

```C
/**
 * Set the address of an existing server of the cluster that this server should join.
```

or any equivalent wording.

Note that any address of an existing server in the cluster is fine and that the address of the bootstrap node is not needed for joining (the bootstrap node might not even exist anymore, if it was removed from the cluster).

Reply:
I mean, the address of the bootstrap node is not needed specifically. Of course you could pass it to `dqlite_server_set_peer_address`, but you don't need to know that it s the bootstrap server, and also if you don't pass it at all, any other address of an existing server is fine. Hope it's clear what I mean.